source/utilities.cc

```diff
@@ -2535,7 +2535,67 @@ namespace aspect
       return operator_list;
     }
 
+    template <int dim>
+    GeneralizedPoint<dim>::GeneralizedPoint(Point<dim> position, const GeometryModel::Interface<dim> *geometry_model)
+    {
+      cartesian_position = position;
+      coordinate_system = geometry_model->natural_coordinate_system();
+      coordinates = geometry_model->cartesian_to_natural_coordinates(cartesian_position);
+    }
+
+    template <int dim>
+    Point<dim> &GeneralizedPoint<dim>::get_position()
+    {
+      return cartesian_position;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim> &GeneralizedPoint<dim>::get_coordinates()
+    {
+      return coordinates;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim-1> GeneralizedPoint<dim>::get_surface_coordinates() const
+    {
+      if (coordinate_system == Coordinates::CoordinateSystem::cartesian)
+        {
+          if (dim == 2)
+            {
+              std_cxx11::array<double,1> a {coordinates[0]};
+              return reinterpret_cast<std_cxx11::array<double,dim-1>& >(a);
```

Comment:
this is a bit ugly but I assume ``return a;`` doesn't compile? I think it would make sense to use template specialization and implement this function for each dimension separately.

Reply:
Yes, I agree. I think that would make things simpler.